### PR TITLE
Add ESM builds to devextreme-react

### PR DESCRIPTION
**Pull request by GoodDayForSurf**

### CODE REVIEW

File: packages/devextreme-react/gulpfile.js [hunk 1]
```diff
@@ -65,15 +70,55 @@ gulp.task(NPM_README, gulp.series(
...
         config.src,
         `!${config.testSrc}`
     ])
         .pipe(ts('tsconfig.json'))
-        .pipe(gulp.dest(config.npm.dist))
+        .pipe(gulp.dest(config.npm.dist + '/cjs'))
+));
+
+gulp.task(NPM_PREPARE_FOLDERS, (done) => {
+    makeModuleForFolder('common');
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js]: Seems that we can create const array with folders and then use forEach to call makeModuleForFolder for them

File: packages/devextreme-react/gulpfile.js [hunk 2]
```diff
@@ -111,5 +156,73 @@ gulp.task(NPM_BUILD_WITH_HEADERS, gulp.series(
...
+    try {
+        fs.mkdirSync(path.join(distFolder, folderPath));
+
+        if (fs.existsSync(path.join(distFolder, 'esm', folderPath, 'index.js'))) {
+            generatePackageJsonFile(folderPath);
+        }
+
+        moduleFileNames.forEach((moduleName) => {
+            makeModuleForFile(moduleName, folderPath);
+        })
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js:185]: Add semicolon

File: packages/devextreme-react/gulpfile.js [hunk 3]
```diff
@@ -111,5 +156,73 @@ gulp.task(NPM_BUILD_WITH_HEADERS, gulp.series(
...
+    );
+
+    const relativeBase = '../'.repeat(relativePath.split('..').length - 1);
+
+    fs.writeFileSync(path.join(absoluteModulePath , 'package.json'), JSON.stringify({
+        sideEffects: false,
+        main:  `${relativeBase}cjs/${moduleFilePath}.js`,
+        module: `${relativeBase}esm/${moduleFilePath}.js`,
+        typings: `${relativeBase}cjs/${moduleFilePath}.d.ts`,
+    },null, 2))
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js]: Add semicolon and space before null

File: packages/devextreme-react/gulpfile.js [hunk 4]
```diff
@@ -111,5 +156,73 @@ gulp.task(NPM_BUILD_WITH_HEADERS, gulp.series(
...
+        NPM_PREPARE_MODULES,
+        NPM_PREPARE_FOLDERS,
+    ),
     shell.task(['npm pack'], {cwd: config.npm.dist})
 ));
+
+function makeModuleForFolder(folderPath, moduleFileNames) {
+    const distFolder = path.join(__dirname, config.npm.dist);
+
+    moduleFileNames = moduleFileNames || findJsModuleFileNamesInFolder(path.join(distFolder, 'esm', folderPath));
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js]: it is not ok that you overwrite argument

File: packages/devextreme-react/gulpfile.js [hunk 5]
```diff
@@ -111,5 +156,73 @@ gulp.task(NPM_BUILD_WITH_HEADERS, gulp.series(
...
+}
+
+function makeModuleForFile(moduleName, folder = null) {
+    fs.mkdirSync(path.join(__dirname, config.npm.dist, folder || '', moduleName));
+
+    generatePackageJsonFile(folder, moduleName);
+}
+
+function generatePackageJsonFile(folder, moduleName) {
+    moduleName = moduleName || '';
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js]: don't overwrite the argument

File: packages/devextreme-react/gulpfile.js [hunk 6]
```diff
@@ -111,5 +156,73 @@ gulp.task(NPM_BUILD_WITH_HEADERS, gulp.series(
...
+        module: `${relativeBase}esm/${moduleFilePath}.js`,
+        typings: `${relativeBase}cjs/${moduleFilePath}.d.ts`,
+    },null, 2))
+}
+
+function findJsModuleFileNamesInFolder(dir) {
+    let results = [];
+    const files = fs.readdirSync(dir);
+
+    for (const file of files) {
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js]: seems that we can use files.filter((file) => ....

File: packages/devextreme-react/gulpfile.js [hunk 7]
```diff
@@ -111,5 +156,73 @@ gulp.task(NPM_BUILD_WITH_HEADERS, gulp.series(
...
+        moduleFileNames.forEach((moduleName) => {
+            makeModuleForFile(moduleName, folderPath);
+        })
+    } catch (error) {
+        error.message = `Exception while makeModuleForFolder(${folderPath}).\n ${error.message}`;
+        throw(error);
+    }
+}
+
+function makeModuleForFile(moduleName, folder = null) {
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js]: Use folder = '' as default

File: packages/devextreme-react/gulpfile.js [hunk 8]
```diff
@@ -111,5 +156,73 @@ gulp.task(NPM_BUILD_WITH_HEADERS, gulp.series(
...
+
+function makeModuleForFolder(folderPath, moduleFileNames) {
+    const distFolder = path.join(__dirname, config.npm.dist);
+
+    moduleFileNames = moduleFileNames || findJsModuleFileNamesInFolder(path.join(distFolder, 'esm', folderPath));
+
+    try {
+        fs.mkdirSync(path.join(distFolder, folderPath));
+
+        if (fs.existsSync(path.join(distFolder, 'esm', folderPath, 'index.js'))) {
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js]: You already created path path.join(distFolder, 'esm', folderPath) above

File: packages/devextreme-react/gulpfile.js [hunk 9]
```diff
@@ -111,5 +156,73 @@ gulp.task(NPM_BUILD_WITH_HEADERS, gulp.series(
...
+function makeModuleForFile(moduleName, folder = null) {
+    fs.mkdirSync(path.join(__dirname, config.npm.dist, folder || '', moduleName));
+
+    generatePackageJsonFile(folder, moduleName);
+}
+
+function generatePackageJsonFile(folder, moduleName) {
+    moduleName = moduleName || '';
+    const absoluteModulePath = path.join(__dirname, config.npm.dist, folder || '', moduleName);
+    const moduleFilePath = (folder ? folder + '/' : '') + (moduleName || 'index');
```
Comment (dxvladislavvolkov) [inline packages/devextreme-react/gulpfile.js]: Use path.join to create path. Also, seems that you can move it to [199 line](https://github.com/DevExpress/DevExtreme/pull/25483/files#diff-ba5fd6cb7491c97328544cdc73f6d49547ff0ecc3ae4d2b06da2dd2cd19b3c43R199)

Comment (GoodDayForSurf) [reply packages/devextreme-react/gulpfile.js]: path.join uses OS specific delimiter but we need only '/' for package.json

File: packages/devextreme-react/tsconfig.esm.json [hunk 10]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (ivanblinov2k17) [inline packages/devextreme-react/tsconfig.esm.json:1]: Why do we need a separate tsconfig for esm?

**Review by ivanblinov2k17 — DISMISSED** [review]

**Review by ivanblinov2k17 — APPROVED** [review]

File: packages/devextreme-react/tsconfig.esm.json (hunk 10 above)
Comment (GoodDayForSurf) [reply packages/devextreme-react/tsconfig.esm.json:1]: in default tsconfig. we compile to cjs, in tsconfig.esm we force compile to build esm in it, also disable generating d.ts for /esm folder

**Review by dxvladislavvolkov — APPROVED** [review]

**Pull request merged**